host/cli/fix.go

```diff
@@ -0,0 +1,607 @@
+
+	if err := f.fixDiscoverd(); err != nil {
+		return err
+	}
+	if err := f.fixFlannel(); err != nil {
+		return err
+	}
+
+	f.l.Info("waiting for discoverd to be available")
+	timeout := time.After(time.Minute)
+	for {
+		var err error
+		if _, err = discoverd.GetInstances("discoverd", 30*time.Second); err != nil {
+			time.Sleep(100 * time.Millisecond)
+		} else {
+			break
+		}
+		select {
+		case <-timeout:
+			return fmt.Errorf("timed out waiting for discoverd, last error: %s", err)
+		}
+	}
+
+	f.l.Info("checking for running controller API")
+	controllerService := discoverd.NewService("controller")
+	controllerInstances, _ := controllerService.Instances()
+	if len(controllerInstances) > 0 {
+		f.l.Info("found running controller API instances", "n", len(controllerInstances))
+		if err := f.fixController(controllerInstances, false); err != nil {
+			f.l.Error("error fixing controller", "err", err)
+			// if unable to write correct formations, we need to kill the scheduler so that the rest of this works
+			if err := f.killSchedulers(); err != nil {
+				return err
+			}
+		}
+	}
+
+	if err := f.fixPostgres(); err != nil {
+		return err
+	}
+
+	f.l.Info("checking for running controller API")
+	controllerInstances, _ = controllerService.Instances()
+	if len(controllerInstances) == 0 {
+		// kill schedulers to prevent interference
+		if err := f.killSchedulers(); err != nil {
+			return err
+		}
+		controllerInstances, err = f.startAppJob("controller", "web", "controller")
+		if err != nil {
+			return err
+		}
+	} else {
+		f.l.Info("found running controller API instances", "n", len(controllerInstances))
+	}
+
+	if err := f.fixController(controllerInstances, true); err != nil {
+		f.l.Error("error fixing controller", "err", err)
+		return err
+	}
+
+	f.l.Info("cluster fix complete")
+
+	return nil
+}
+
+func (f *clusterFixer) startAppJob(app, typ, service string) ([]*discoverd.Instance, error) {
+	f.l.Info(fmt.Sprintf("no %s %s process running, getting release details from hosts", app, typ))
+	releases := f.findAppReleaseJobs(app, typ)
+	if len(releases) == 0 {
+		return nil, fmt.Errorf("didn't find any %s %s release jobs", app, typ)
+	}
+
+	// get a job template from the first release
+	var job *host.Job
+	for _, job = range releases[0] {
+		break
+	}
+	job.ID = cluster.GenerateJobID(f.hosts[0].ID())
+	f.fixJobEnv(job)
+	// run it on a host
+	f.l.Info(fmt.Sprintf("starting %s %s job", app, typ), "job.id", job.ID, "release", job.Metadata["flynn-controller.release"])
+	if err := f.hosts[0].AddJob(job); err != nil {
+		return nil, fmt.Errorf("error starting %s %s job: %s", app, typ, err)
+	}
+	f.l.Info("waiting for job to start")
+	return discoverd.GetInstances(service, time.Minute)
+}
+
+func (f *clusterFixer) fixDiscoverd() error {
+	f.l.Info("ensuring discoverd is running on all hosts")
+	releases := f.findAppReleaseJobs("discoverd", "app")
+	if len(releases) == 0 {
+		return fmt.Errorf("didn't find any discoverd release jobs")
```

Comment:
This returns an error on a freshly booted cluster as there are no discoverd jobs in either the [done or crashed state](https://github.com/flynn/flynn/blob/first-aid/host/cli/fix.go#L563-L565) (I was trying to use `flynn-host fix` to resurrect a scheduler that I killed)